File: src/rails/game/model/BaseTokensModel.java
package rails.game.model;

import java.util.Set;

import com.google.common.collect.ImmutableSet;
import com.google.common.collect.ImmutableSortedSet;
import com.google.common.collect.Iterables;
import com.google.common.collect.Sets;

import rails.game.BaseToken;
import rails.game.PublicCompany;
import rails.game.state.Model;
import rails.game.state.Portfolio;
import rails.game.state.PortfolioSet;

/**
 * A model presenting the number of tokens
 */
public class BaseTokensModel extends Model {

    // the free tokens belong to the company
    private final PortfolioSet<BaseToken> freeBaseTokens;
    // a list of all base tokens, configured later
    private ImmutableSortedSet<BaseToken> allTokens;

    private BaseTokensModel(PublicCompany parent, String id) {
        super(parent, id);
        freeBaseTokens = PortfolioSet.create(parent, "freeBaseTokens", BaseToken.class);
    }

    public static BaseTokensModel create(PublicCompany parent, String id){
        return new BaseTokensModel(parent, id);
    }

    /**
     * Initialize a set of tokens
     */
    public void initTokens(Set<BaseToken> tokens) {
        allTokens = ImmutableSortedSet.copyOf(tokens);
        Portfolio.moveAll(allTokens, getParent());
    }
    
    /**
     * @return parent the public company
     */
    @Override
    public PublicCompany getParent() {
        return (PublicCompany)super.getParent();
    }
    
    /**
     * @return the next (free) token to lay, null if none is available
     */
    public BaseToken getNextToken() {
        if (freeBaseTokens.size() == 0) return null;
        return Iterables.get(freeBaseTokens, 0);
    }
    
    public ImmutableSet<BaseToken> getAllTokens() {
        return allTokens;
    }
    
    public ImmutableSet<BaseToken> getFreeTokens() {
        return freeBaseTokens.items();
    }
    
    public ImmutableSet<BaseToken> getLaidTokens() {
        return Sets.difference(allTokens, freeBaseTokens.items()).immutableCopy();
    }
    
    public int nbAllTokens() {
        return allTokens.size();
    }
    
    public int nbFreeTokens() {
        return freeBaseTokens.size();
    }
    
    public int nbLaidTokens() {
        return allTokens.size() - freeBaseTokens.size();
    }
    
    /**
     * @return true if token is laid
     */
    public boolean tokenIsLaid(BaseToken token) {
        return allTokens.contains(token);
    }
    
    @Override 
    public String toText() {
        int allTokens = nbAllTokens();
        int freeTokens = nbFreeTokens();
        if (allTokens == 0) {
            return "";
        } else if (freeTokens == 0) {
            return "-/" + allTokens;
        } else {
            return freeTokens + "/" + allTokens;
        }
    }

}

File: dotnet/RailsLib.Net/Net/Game/Model/BaseTokensModel.cs
using GameLib.Net.Game.State;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Text;

/**
 * A model presenting the number of tokens
 */

namespace GameLib.Net.Game.Model
{
    public class BaseTokensModel : RailsModel
    {
        // the free tokens belong to the company
        private PortfolioSet<BaseToken> freeBaseTokens;
        // a list of all base tokens, configured later
        private SortedSet<BaseToken> allTokens;

        private BaseTokensModel(PublicCompany parent, string id) : base(parent, id)
        {
            freeBaseTokens = PortfolioSet<BaseToken>.Create(parent, "freeBaseTokens");
            freeBaseTokens.AddModel(this);
        }

        public static BaseTokensModel Create(PublicCompany parent, string id)
        {
            return new BaseTokensModel(parent, id);
        }

        /**
         * Initialize a set of tokens
         */
        public void InitTokens(IEnumerable<BaseToken> tokens)
        {
            allTokens = new SortedSet<BaseToken>(tokens);// ImmutableSortedSet.copyOf(tokens);
            Portfolio.MoveAll(allTokens, Parent);
        }

        /**
         * @return parent the public company
         */
        new public PublicCompany Parent
        {
            get
            {
                return (PublicCompany)base.Parent;
            }
        }

        /**
         * @return the next (free) token to lay, null if none is available
         */
        public BaseToken GetNextToken()
        {
            if (freeBaseTokens.Count == 0) return null;
            var i = freeBaseTokens.GetEnumerator();
            i.MoveNext();
            return i.Current;
            //return Iterables.get(freeBaseTokens, 0);
        }

        public IReadOnlyCollection<BaseToken> GetAllTokens()
        {
            return allTokens;
        }

        public IReadOnlyCollection<BaseToken> GetFreeTokens()
        {
            return freeBaseTokens.Items;
        }

        public IReadOnlyCollection<BaseToken> GetLaidTokens()
        {
            var diff = new SortedSet<BaseToken>(allTokens);
            diff.ExceptWith(freeBaseTokens.Items);
            return diff;
            //return Sets.difference(allTokens, freeBaseTokens.items()).immutableCopy();
        }

        public int NbAllTokens
        {
            get
            {
                return allTokens.Count;
            }
        }

        public int NbFreeTokens
        {
            get
            {
                return freeBaseTokens.Count;
            }
        }

        public int NbLaidTokens
        {
            get
            {
                return allTokens.Count - freeBaseTokens.Count;
            }
        }

        /**
         * @return true if token is laid
         */
        public bool TokenIsLaid(BaseToken token)
        {
            return allTokens.Contains(token);
        }

        override public string ToText()
        {
            int allTokens = NbAllTokens;
            int freeTokens = NbFreeTokens;
            if (allTokens == 0)
            {
                return "";
            }
            else if (freeTokens == 0)
            {
                return "-/" + allTokens;
            }
            else
            {
                return freeTokens + "/" + allTokens;
            }
        }
    }
}
